### ISSUE TITLE
FEA Discrete lift curve + scores implementation

### ISSUE THREAD

**Issue by vitaliset** [dihanster/datalib#17]
What I'm calling "discrete_lift_curve": I simply sort by score and see how the average proportion of score groups behave. Ideally groups with lower scores should also have lower average `y` (both in the case of predict_proba for classification and in the case of predict for regression).

I also have some some statistics that I can get from this curve such as the mean difference between each bin. This can be interpreted as scores.

I have this python script that implements the metric, but I'm using pandas. Probably chatGPT can help us change the pandas to numpy.
Also, this implementation uses "mean" as aggregation for the "groupby" but I think we should let the user define what he wants (sometimes a quantile will be more valuable).

```python
import numpy as np
import pandas as pd

from sklearn.utils.validation import _check_sample_weight
from pandas._libs.lib import is_integer

def weighted_qcut(values, weights, q, **kwargs):
    # https://stackoverflow.com/questions/45528029/python-how-to-create-weighted-quantiles-in-pandas
    'Return weighted quantile cuts from a given series, values.'
    if is_integer(q):
        quantiles = np.linspace(0, 1, q + 1)
    else:
        quantiles = q
    order = pd.Series(weights).iloc[values.argsort()].cumsum()
    bins = pd.cut(order / order.iloc[-1], quantiles, **kwargs)
    return bins.sort_index()#.values

def discrete_lift_curve(y_true, y_pred, bins=10, sample_weight=None):
    sample_weight = _check_sample_weight(sample_weight, y_true)
    bins_scores = weighted_qcut(y_pred, sample_weight, bins, labels=False)
    return (np.array(list(range(bins))),
            np.array([np.average(y_true[bins_scores==q], weights=sample_weight[bins_scores==q]) for q in range(bins)]))
     
def last_bin_of_discrete_lift_curve(y_true, y_pred, bins=10, sample_weight=None):
    _, values = discrete_lift_curve(y_true, y_pred, bins=bins, sample_weight=sample_weight)
    return values[-1]
 
def mean_diff_of_discrete_lift_curve(y_true, y_pred, bins=10, sample_weight=None):
    _, values = discrete_lift_curve(y_true, y_pred, bins=bins, sample_weight=sample_weight)
    return pd.Series(values).diff().mean()
```

Should this be in this library?
